types/wlr_xdg_foreign_v1.c

```diff
@@ -0,0 +1,437 @@
+			struct wlr_xdg_surface_v6 *xdg_surface_v6 =
+				wlr_xdg_surface_v6_from_wlr_surface(surface);
+			struct wlr_xdg_surface_v6 *xdg_child_v6 =
+				wlr_xdg_surface_v6_from_wlr_surface(child);
+			wlr_xdg_toplevel_v6_set_parent(xdg_child_v6, xdg_surface_v6);
+			wl_signal_add(&xdg_child_v6->events.unmap, &imported->child_unmap);
+		}
+	}
+}
+
+static const struct zxdg_imported_v1_interface xdg_imported_impl = {
+	.destroy = xdg_imported_handle_destroy,
+	.set_parent_of = xdg_imported_handle_set_parent_of
+};
+
+static struct wlr_xdg_exported_v1 *xdg_exported_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exported_v1_interface,
+		&xdg_exported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_exported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zxdg_exported_v1_interface xdg_exported_impl = {
+	.destroy = xdg_exported_handle_destroy
+};
+
+static void xdg_exporter_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wlr_xdg_exporter_v1 *xdg_exporter_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_exporter_v1_interface,
+		&xdg_exporter_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_exported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_exported_v1 *exported =
+		xdg_exported_from_resource(resource);
+	struct wlr_xdg_imported_v1 *imported, *imported_tmp;
+	wl_list_for_each_safe(imported, imported_tmp,
+			&exported->imports, export_link) {
+		zxdg_imported_v1_send_destroyed(imported->resource);
+		wl_resource_destroy(imported->resource);
+	}
+	wl_list_remove(&exported->xdg_surface_unmap.link);
+	wl_list_remove(&exported->link);
+	free(exported->handle);
+	free(exported);
+}
+
+static void handle_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_exported_v1 *exported =
+		wl_container_of(listener, exported, xdg_surface_unmap);
+	wl_resource_destroy(exported->resource);
+}
+
+static void xdg_exporter_handle_export_toplevel(struct wl_client *wl_client,
+		struct wl_resource *client_resource,
+		uint32_t id,
+		struct wl_resource *surface_resource) {
+	struct wlr_xdg_exporter_v1 *exporter =
+		xdg_exporter_from_resource(client_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
+
+	if (!verify_is_toplevel(client_resource, surface)) {
+		return;
+	}
+
+	struct wlr_xdg_exported_v1 *exported =
+		calloc(1, sizeof(struct wlr_xdg_exported_v1));
+	if (exported == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	exported->surface = surface;
+	exported->resource = wl_resource_create(wl_client,
+		&zxdg_exported_v1_interface, wl_resource_get_version(exporter->resource),
+		id);
+	exported->handle = calloc(11, sizeof(char));
```

Comment:
Why not use `sizeof(exported) + 3`?

Reply:
`sizeof(char)` is guaranteed to be 1.

Reply:
I was speaking about the 11

Reply:
I know, there are two problems on this line.

Reply:
oh right :)

Reply:
Sorry I should have left a comment. This code is a placeholder and will be removed.

Reply:
Replaced with the UUID generation shim.